DataManagementSystem/DB/FileCatalogComponents/MultiVOFileMetadata.py

```diff
@@ -0,0 +1,17 @@
+""" DIRAC Multi VO FileCatalog plugin class to manage file metadata for multiple VO.
+"""
+
+__RCSID__ = "$Id$"
+
+from DIRAC.DataManagementSystem.DB.FileCatalogComponents.FileMetadata import FileMetadata
+from DIRAC.DataManagementSystem.DB.FileCatalogComponents.MultiVOMetaNameMixIn import MultiVOMetaNameMixIn
+
+
+class MultiVOFileMetadata(MultiVOMetaNameMixIn, FileMetadata):
+  """
+  MULti-VO FileCatalog plugin implementation.
+  """
+
+  def __init__(self, database=None):
+    FileMetadata.__init__(self, database=database)
+    MultiVOMetaNameMixIn.__init__(self)
```

Comment:
These inheritances are partly broken. You are in this situation right now:

```
In [1]: class MetaNameMixInNoObject():
   ...:   def __init__(self):
   ...:     print("I am MetaNameMixInNoObject")
   ...: 
   ...: 
   ...: class MultiVOMetaNameMixInNoObject(MetaNameMixInNoObject):
   ...:   def __init__(self):
   ...:     print("I am MultiVOMetaNameMixInNoObject")
   ...: 
   ...: 
   ...: class DirectoryMetadataNoObject(MetaNameMixInNoObject):
   ...:  def __init__(self):
   ...:   print("I am DirectoryMetadataNoObject")
   ...: 
   ...: 
   ...: class MultiVODirectoryMetadataNoObject(MultiVOMetaNameMixInNoObject, DirectoryMetadataNoObject):
   ...:   def __init__(self):
   ...:     print("I am MultiVODirectoryMetadataNoObject")
   ...:     DirectoryMetadataNoObject.__init__(self)
   ...:     MultiVOMetaNameMixInNoObject.__init__(self)
   ...: 
   ...: 
   ...: MultiVODirectoryMetadataNoObject()
   ...: 
I am MultiVODirectoryMetadataNoObject
I am DirectoryMetadataNoObject
I am MultiVOMetaNameMixInNoObject
```
As you see what's in `MetaNameMixInNoObject` is not executed. This is because you are using old style classes: http://python-history.blogspot.com/2010/06/method-resolution-order.html

This might not be a problem with the current code because there's nothing really done in  `MetaNameMixInNoObject`, but we shouldn't rely on that.


What should instead be done is the following (inheriting from `object` and using `super`:
```
In [1]: class MetaNameMixIn(object):
   ...:   def __init__(self):
   ...:     print("I am MetaNameMixIn")
   ...: 
   ...: 
   ...: class MultiVOMetaNameMixIn(MetaNameMixIn):
   ...:   def __init__(self):
   ...:     super(MultiVOMetaNameMixIn, self).__init__()
   ...:     print("I am MultiVOMetaNameMix")
   ...: 
   ...: 
   ...: class DirectoryMetadata(MetaNameMixIn):
   ...:   def __init__(self):
   ...:     super(DirectoryMetadata, self).__init__()
   ...:     print("I am DirectoryMetadata")
   ...: 

In [2]: class MultiVODirectoryMetadataWithSuper(MultiVOMetaNameMixIn, DirectoryMetadata):
   ...:   def __init__(self):
   ...:     super(MultiVODirectoryMetadataWithSuper, self).__init__()
   ...:     print("I am MultiVODirectoryMetadataWithSuper")
   ...: 

In [3]: MultiVODirectoryMetadataWithSuper()
I am MetaNameMixIn
I am DirectoryMetadata
I am MultiVOMetaNameMix
I am MultiVODirectoryMetadataWithSuper
```

Also, python 3 by default would use the MRO of new-style classes.

Reply:
Sorry for not replying straight away, but I was on holidays and I'm taking some days off next week again ...

I totally agree I should be using new style classes, I'll fix this.  However your example above is not quite accurate since I explicitly call base class inits:
https://github.com/DIRACGrid/DIRAC/blob/0d83eddd103d050edcec6abaf58b2ea32f3ba532/DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py#L16-L19

and also:

https://github.com/DIRACGrid/DIRAC/blob/0d83eddd103d050edcec6abaf58b2ea32f3ba532/DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py#L15-L16

Another problem with `super` is that all `init` prototypes have to agree, and in our example we have an additional `database=`  argument in one of them. I could pass it as a default to another one but it would not look nice.

How do you see this ?

   best, JM

Reply:
You can define `def __init__(self, *args, **kwargs):` for absorbing all the arguments.

Reply:
Hi, the inheritance tree has been updated.